### DISCUSSION TITLE
Infer type of searchParams

### DISCUSSION THREAD

**Discussion by missile-developer** [47ng/nuqs#571] [Q&A] [answered]
I want to refer my fetcher query params, In Zod you can use z.infer<typeof searchParams>. How about nuqs?

**Comment by franky47** [comment]
Assuming you have an object describing your search params as such:

```ts
import { parseAsInteger, parseAsBoolean, parseAsString } from 'nuqs'

const searchParams = {
  int: parseAsInteger,
  bool: parseAsBoolean,
  stringWithDefault: parseAsString.withDefault('')
}

// And you want the following type as output?
type inferred = {
  int: number | null
  bool: boolean | null
  stringWithDefault: string
}
```

There is a (contrived) way to do this currently, but it could definitely use a little abstraction, I'll have a look at what I can do.

```ts
import type { createSearchParamsCache } from 'nuqs/server'

type inferred = ReturnType<
  ReturnType<typeof createSearchParamsCache<typeof searchParams>>['parse']
>
```

**Reply by franky47** [reply]
Edit: here's a better abstraction, let me know if it works for you and I can add it to the types API.

```ts
export type inferParserType<Parser> = Parser extends ParserBuilder<
  infer Type
> & {
  defaultValue: infer Type
}
  ? Type
  : Parser extends ParserBuilder<infer Type>
    ? Type | null
    : never

export type inferParserRecordType<
  Map extends Record<string, ParserBuilder<any>>
> = {
  [Key in keyof Map]: inferParserType<Map[Key]>
}

// Usage:
type searchParamsType = inferParsersType<typeof searchParams>
```

**Reply by franky47** [reply]
Implementation being worked on in #578.

**Reply by wottpal** [reply]
Doesn't really work for me:

![CleanShot 2024-08-30 at 08 28 29@2x](https://github.com/user-attachments/assets/308df720-6880-42d0-8ef3-f9e8c257e11a)

Would also love to see some kind of type inference being merged as well. 🫶

**Reply by franky47** [reply]
@wottpal you'll need to use `inferParserRecordType` if you are passing it an object/record of keys and associated parsers.

I could try and make it a single inference type that tries to detect whether it's working on a parser or a record of parsers, but that would be a wild ride.

**Reply by franky47** [reply]
Actually it wasn't that bad, leveraging the two inference helpers defined before, we can now have a single `inferParserType` that can handle both single parsers or objects/records of key/parser pairs 🎉 

https://github.com/47ng/nuqs/blob/c2b490885e8b5d351e63e39761ac499a6c6b05bb/packages/nuqs/src/parsers.ts#L446-L451

I'll get this merged ASAP.

**Reply by franky47** [reply, marked as answer]
Shipped in version 1.18.0 🚀 

https://nuqs.47ng.com/docs/utilities#parser-type-inference

**Reply by wottpal** [reply]
Works like a breeze. 🫶
